zkevm-circuits/src/util/cell_manager.rs

```diff
@@ -280,4 +360,8 @@ impl<Stats, S: CellManagerStrategy<Stats = Stats>> CellManager<S> {
     pub fn get_stats(&self) -> Stats {
         self.strategy.get_stats(&self.columns)
     }
+
+    pub fn get_strategy(&mut self) -> &mut S {
+        &mut self.strategy
+    }
```

Comment:
I wouldn't do (1) but because in my opinion this feature looks like a keccak specific thing.

Regarding (2), isn't bending the meaning of the trait that cell_manager->deref_mut returns a strategy instead of a cell manager?

Reply:
For (2), I think deref allows you to call both methods of the cell_manager and methods of the strategy.

But I agree with you on point (1) and that (2) might be confusing. The status quo looks good to me now.

Reply:
Feel that, if there are many code duplication, then it might be some place to improve.

Notice that, the CellManager is like a just columns warehouse
```
pub(crate) struct CellManager<S: CellManagerStrategy> {
    columns: CellManagerColumns,
    strategy: S,
}
```

and most of columns manipulation logic are now based on strategy S. That's why there are a lot of `bypass logic` on cell manager because we try to call a method on column warehouse. 

How about we swap the role so the strategy is new "cellmanager", then during initialization we just pass a column warehouse to it? Feel it can save lot of code re-implementation

Reply:
So, precisely the point of this is to avoid each of the modules to use a different cell manager directly because they will start to diverge, even if we say that they have to use the same trait.

This is the strategy pattern https://en.wikipedia.org/wiki/Strategy_pattern

All users will only communicate with the same struct Cell Manager, the parts that are different behaviour (how cells are placed) will be handle by individual strategies that will implement different behaviours.

This was presented as the design of this feature: https://hackmd.io/_RIuqSZSTp-p8fZ4ohbyVg which was available for feedback for months.

However, for a bad design (from the software engineering POV) of the keccak circuit had made necessary to add some weird stuff like the value queries. Keccak queries the cells all over in functions and such.

The right design for a circuit using the Cell Manager would be to always declare the signals before using them, and from that declaration used them both in the setup and the witness generation. This allow simpler Cell Manager code and more clever optimizations. Like it is done in Chiquito: https://github.com/privacy-scaling-explorations/chiquito/blob/main/src/compiler/cell_manager.rs

I hope this PR help us to get in that direction.

Reply:
Thanks for the explanation!

A essential different I spotted is, in strategy_patten, `class`, in our case, cellmamager, do not need to implement strategy interface, and class is more like a high level encapsulation of its strategy underlyne. It's functionality shouldn't be one to one mapping. Take [Wi](https://en.wikipedia.org/wiki/Strategy_pattern) example, a `car` is a class, where one of its functionality `brake` will be implemented as different strategy. However, car do not implement brake interface.

In cell manager case, its more like we have a unnessary abstraction to have a `Car` encaptulate another `Car`. From this perspective, does it make sense that every modules just rely on CellManagerStrategy trait, where 2 implementation column/row oriented for it ? (CellManagerStrategy can just rename to CellManager)

Reply:
1.

The design was available months ago. It was encouraged in the scrum meetings to go and comment, and it was approved by Edu and CC. Adría and Carlos also checked it.

https://hackmd.io/_RIuqSZSTp-p8fZ4ohbyVg  

https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1203#issuecomment-1516197888

https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1203#issuecomment-1516384404

2.

Following your comments I have changed some naming in the code to make more clear the intention. I also have moved out of the strategy the only repeated code, the creation of the Cell.

Additionally, I have added comments to clarify that query_value is deprecated. The configuration and synthesize should share the cells in the same way they share the columns.

3.

If we do what you propose, that would return to the previous status quo, where each circuit has a different cell manager.

4.

Because of the example in the wikipedia you might have inferred that the “Context class” (Car) should implement more features than the strategy interface (IBrake). That is never stated as the intention of the Strategy Pattern. It is just a choice of example.

5.

In this more extended article https://refactoring.guru/design-patterns/strategy

We can see that in its example, the context class (Navigator) only goal from the point of the client is to build a route, and then it wraps different ways of doing it (walking, by car…) in different strategies that implement the same interface.

It is not necessary	or condition or related at all that the context class have other actions beyond the one done in the strategies. It is an interface between the client and the strategies.

6.

The goal of the strategy pattern is NOT to wrap different behaviours in a single class with internal implementer for each behaviour, like Break, Accelerate, etc,,

The goal of the strategy pattern is to give the clients of a problem solver a single point of entry to different strategies to solve that problem.

So in our cases, the point is that all circuits will communicate with only one struct CellManager, and then the strategy used will be transparent to the circuit.

Reply:
@hero78119 I think @leolara has a point on the strategy design.
Can we maintain the current scope of the PR and move the potential design change to a new issue and PR?